### ISSUE TITLE
Problem while initializing the network with same set of weights?

### ISSUE THREAD

**Issue by ajeetksingh** [facebook/fb.resnet.torch#56]
Hi,

I am doing these runs, where I have to initialize the network with same set of weights.
In this case, the training and testing errors should be similar if not the same.
But, I am getting different training and testing errors in each and every run.

IMO, this disparity can be caused by following two things:
a) Initializing the weights with different values on every run.
    --  To overcome it, I have written the weights in a file, which I am loading on every run.
b) Randomness in the way we present the data to the network.
    -- Data Shuffling: I have disabled it by providing the manualSeed.
    -- Data preprocessing: I am only doing the color normalization.
    -- I have also disabled the randomCrop which is by default 'false'.

Here is the code snippet, I have added to the 'models/resnet.lua' to write and load the weights initialized.

`--a = v.weight:normal(0,math.sqrt(2/n))`
` filename = 'randwts/' .. name .. '_' .. k .. '_init_wts.t7'`
`--torch.save(filename, a)`
` --a = torch.load(filename)`
`v.weight = torch.load(filename)`

What's the step which I am missing here? 

Thanks.

**Comment by colesbury** [comment]
Why would you expect the training and test error to be the same?

**Comment by ajeetksingh** [comment]
@colesbury 
I think, I was not able to convey the question properly.
I do not expect the training and test error to be the same. 

But for different runs, with the network initialized with the same set of weights, I would expect similar training errors, with no randomness, whatsoever, in the way I present the data to the network. For example, 

Run 1 - Training Error is x%; Testing Error is a%
Run 2 - Training Error is y%; Testing Error is b%.
So, I am presenting the data to the network in the same order, without any randomness. I am initializing the network with the same set of weights in both the runs. Hence, logically, I would
expect the 
`x = y` and `a = b`
Correct? But somehow, I am not getting that. Is there any other piece of code which I have to change, the way I did here to stop the shuffling of data?

https://github.com/facebook/fb.resnet.torch/blob/65e675a8cc7c96918d140656af55941afbfa8104/dataloader.lua#L65

Thanks.

**Comment by soumith** [comment]
not correct. The code to do certain operations in the network is not deterministic. It will give slightly different results every time you pass the same input. 

Read this thread for example: https://groups.google.com/forum/#!searchin/torch7/deterministic/torch7/bkVMKmmwG1A/_65nkl5aDQAJ

**Comment by ajeetksingh** [comment]
OK. That answers my question.

The thread you mention, one guy has mentioned that we can make the 'cudnn' deterministic by setting the mode of modules as Module:setMode(1, 1, 1). Is this possible?

Thanks.

**Comment by soumith** [comment]
give it a try

**Comment by ajeetksingh** [comment]
For the first epoch, at least I am getting same training error for every runs. And for latter epochs, the training errors are slightly different but not much, across runs.

Following thread helped. 
https://github.com/soumith/cudnn.torch/issues/92

Thanks.
